testutil/wallet.go

```diff
@@ -0,0 +1,265 @@
+package testutil
+
+import (
+	"fmt"
+	"sync"
+
+	"go.sia.tech/core/chain"
+	"go.sia.tech/core/types"
+	"go.sia.tech/coreutils/wallet"
+)
+
+// An EphemeralWalletStore is a Store that does not persist its state to disk. It is
+// primarily useful for testing or as a reference implementation.
+type EphemeralWalletStore struct {
+	privateKey types.PrivateKey
+
+	mu          sync.Mutex
+	uncommitted []*chain.ApplyUpdate
+	tip         types.ChainIndex
+	utxos       map[types.Hash256]types.SiacoinElement
+
+	immaturePayoutTransactions map[uint64][]wallet.Transaction // transactions are never removed to simplify reorg handling
+	transactions               []wallet.Transaction
+}
+
+// ProcessChainRevertUpdate implements chain.Subscriber
+func (es *EphemeralWalletStore) ProcessChainRevertUpdate(cru *chain.RevertUpdate) error {
+	es.mu.Lock()
+	defer es.mu.Unlock()
+
+	// if the update is uncommitted, remove it from the uncommitted list
+	if len(es.uncommitted) > 0 && es.uncommitted[len(es.uncommitted)-1].State.Index == cru.State.Index {
+		es.uncommitted = es.uncommitted[:len(es.uncommitted)-1]
+		return nil
+	}
+
+	walletAddress := types.StandardUnlockHash(es.privateKey.PublicKey())
+
+	// revert any siacoin element changes
+	cru.ForEachSiacoinElement(func(se types.SiacoinElement, spent bool) {
+		if se.SiacoinOutput.Address != walletAddress {
+			return
+		}
+
+		if spent {
+			es.utxos[se.ID] = se
+		} else {
+			delete(es.utxos, se.ID)
+		}
+	})
+
+	// remove any transactions that were added in the reverted block
+	filtered := es.transactions[:0]
+	for _, txn := range es.transactions {
+		if txn.Index == cru.State.Index {
+			continue
+		}
+		filtered = append(filtered, txn)
+	}
+
+	// update element proofs
+	for id, se := range es.utxos {
+		cru.UpdateElementProof(&se.StateElement)
+		es.utxos[id] = se
+	}
+	return nil
+}
+
+// ProcessChainApplyUpdate implements chain.Subscriber
+func (es *EphemeralWalletStore) ProcessChainApplyUpdate(cau *chain.ApplyUpdate, commit bool) error {
+	es.mu.Lock()
+	defer es.mu.Unlock()
+
+	es.uncommitted = append(es.uncommitted, cau)
+	if !commit {
+		return nil
+	}
```

Comment:
this confuses me. Subscribers that don't write anything to persistent storage should never need to buffer updates like this.

Reply:
I copied it mostly to show how a real implementation would need to work

Reply:
A subscriber with a transactional DB wouldn't look like this either, though. You'd write things as usual and then commit the transaction on `mayCommit`. 🤔

Reply:
With many transactional databases, holding transactions open for long periods is bad practice. In SQLite, when a transaction writes, all other threads are blocked until it's committed. In MySQL and Postgres, you will hit table locks. Those stores would require a buffer to cache before opening a transaction.

The best practice is to open a transaction, write everything, and commit or roll back as quickly as possible.